Call mb_init in case of failed locale search, or failed locale lookup to properly init utf-8 encoding.

Breakster opened the pull request:
This might not be the nicest way to fix the utf-8 issues, but it is probably the least invasive and it does the job.

Breakster committed:
Call mb_init in case of failed locale search, or failed locale lookup

justinmk commented:
Some test cases would help speed this through.

Breakster commented:
Ok, I'm not sure how about general builds, but on my system this change fixed the utf8_spec.lua legacy testcase.

Can you advise what else to do?

justinmk commented:
In https://github.com/neovim/neovim/issues/1672#issuecomment-70347946 you mention that a shell variable isn't picked up. Maybe we need a test for that if it is indeed broken.

Also, maybe add your full explanation to the commit message detail?

> mb_init() is supposed to initialize multibyte support, but it was not called after the default for encoding has been set. 

<!--
Just to discuss, here's why I think this was broken: 

- Vim originally always started with latin1. 
- The doc for `mb_init()` says:

```
Called in three cases:
- by main() to initialize (p_enc == NULL)
- by set_init_1() after 'encoding' was set to its default. 
- by do_set() when 'encoding' has been set.
```

- In Vim, the only way to change "the default" is for a user to `set encoding=utf8` in .vimrc. But this doesn't happen until `source_startup_scripts()` wayyy down in `main()`, after `set_init_2()`.
-->

justinmk commented on src/nvim/option.c:
```diff
@@ -2030,7 +2030,13 @@ void set_init_1(void)
       free(p_enc);
       p_enc = save_enc;
+      /* mb_init() failed. Call it again for the saved encoding */
+      mb_init();
```
I think I follow now. If `set_init_1` fails to set the encoding, it sets `p_enc` back to the default encoding, but it doesn't call `mb_init()` to actually *apply the default (utf8) encoding* because it assumes it was already done at startup...?

Would you mind a quick debugging step to see what `enc_locale()` is returning in your case? Here are quick steps:

- start gbd in "TUI" mode: `gdb -tui build/bin/nvim`
- set a breakpoint at line 1987: `break option.c:1987`
- use the `n` command to step over that line
- use the `p` command to see the value of `p`: `p p`
```

What is the value of `p`?

justinmk replied:
I feel much better about what's going on here now. I updated https://github.com/neovim/neovim/pull/2106#issuecomment-77659442 with a description of what I understand to be the root problem.

Your fix is correct :+1: Thanks a lot!

justinmk commented:
Since you're in the area, could you correct the mention of "latin1" on line 1992.

justinmk commented on src/nvim/option.c:
```diff
@@ -2030,7 +2030,13 @@ void set_init_1(void)
     } else {
       free(p_enc);
       p_enc = save_enc;
+      /* mb_init() failed. Call it again for the saved encoding */
```
// mb_init() on the detected encoding; try again on the default (utf8) encoding.

justinmk commented on src/nvim/option.c:
```diff
@@ -2030,7 +2030,13 @@ void set_init_1(void)
     }
+  } else {
+    /* enc_locale() didn't find a useful locale. Call mb_init() with
+     * the default utf-8 encoding */
```
// enc_locale() didn't find a useful locale; try again on the default (utf8).

justinmk commented:
Just tried this, it fixes https://github.com/neovim/neovim/issues/1271! Hooray!

Breakster committed:
Running `:set encoding=utf-8` after startup correctly initializes multibyte.
But mb_init() was _not_ being called in set_init_1() if locale detection
(enc_locale()) failed. This wasn't a problem in Vim because if locale
detection failed then it used latin1, which of course does not require mb_init().
But we default to utf8, so we need to call mb_init() even if locale detection
fails.

Breakster committed:
Merge branch 'master' of https://github.com/Breakster/neovim

Conflicts:
	src/nvim/option.c

Breakster commented:
The following piece of code in `enc_locale()` tries to get the system locale.

```C
# ifdef HAVE_NL_LANGINFO_CODESET
  if ((s = nl_langinfo(CODESET)) == NULL || *s == NUL)
# endif
#  if defined(HAVE_LOCALE_H) || defined(X_LOCALE)
    if ((s = setlocale(LC_CTYPE, NULL)) == NULL || *s == NUL)
#  endif
      if ((s = (char *)os_getenv("LC_ALL")) == NULL || *s == NUL)
        if ((s = (char *)os_getenv("LC_CTYPE")) == NULL || *s == NUL)
          s = (char *)os_getenv("LANG");
```

On my OSX install, `nl_langinfo(CODESET)` returns US_ASCII. As a result the following shell variables are ignored, even though US_ASCII is not recognized by vim as valid encoding. `enc_locale()` then returns US_ASCII and `mb_init()` fails to recognize it.

I couldn't figure out what the use of `nl_langinfo()` is and whether this is intended behaviour.

@justinmk: I updated the comments like you suggested and I'm glad you like the fix :)

justinmk commented:
rebased and merged. Thanks again!

justinmk commented:
@Breakster By the way, if you have a moment it would be appreciated if you could sign the CLA.
